File: backend/Sources/App/Services/RAGPipeline.swift
import Foundation
import Fluent

/// RAG (Retrieval-Augmented Generation) pipeline for Context-Mapper.
///
/// 1. Chunk long documents into overlapping 400-word chunks
/// 2. Embed and upsert to SQLite vector store (BLOB, SIMD search)
/// 3. Query-time retrieval with 0.72 cosine similarity threshold
/// 4. Context-aware proposal generation via Claude Sonnet 4.6
///
/// BUG FIX: Cold-start branch now passes `industry` to ColdStart.getContext()
/// (the old Python backend had an empty context_chunks list on cold-start).
struct RAGPipeline {
    static let shared = RAGPipeline()
    private init() {}

    private let chunkSize    = 400   // words
    private let chunkOverlap = 80    // words (prevents cutting mid-sentence)

    // MARK: - Chunking

    func chunkText(_ text: String) -> [String] {
        let words = text.split(separator: " ", omittingEmptySubsequences: true).map(String.init)
        var chunks: [String] = []
        var i = 0
        while i < words.count {
            let slice = words[i..<min(i + chunkSize, words.count)]
            let chunk = slice.joined(separator: " ")
            chunks.append(chunk)
            i += chunkSize - chunkOverlap
        }
        // Drop tiny trailing chunks (likely incomplete sentences)
        return chunks.filter { $0.count > 50 }
    }

    // MARK: - Ingest

    /// Chunk, embed, and persist a full proposal document.
    func ingestProposal(
        customerId: String,
        proposalId: String,
        text: String,
        metadata: [String: Any],
        db: any Database
    ) async throws -> Int {
        let chunks = chunkText(text)
        guard !chunks.isEmpty else { return 0 }
        let metas = chunks.map { _ -> [String: Any] in
            var m = metadata
            m["type"] = "proposal_chunk"
            m["customer_id"] = customerId
            m["proposal_id"] = proposalId
            return m
        }
        try await LocalVectorStore.shared.upsert(
            customerId: customerId,
            entryType: "proposal_chunk",
            texts: chunks,
            metadatas: metas,
            db: db
        )
        return chunks.count
    }

    /// Embed and persist brand voice examples.
    func ingestBrandVoice(
        customerId: String,
        brandVoiceId: String,
        text: String,
        metadata: [String: Any],
        db: any Database
    ) async throws -> Int {
        let chunks = chunkText(text)
        guard !chunks.isEmpty else { return 0 }
        let metas = chunks.map { _ -> [String: Any] in
            var m = metadata
            m["type"] = "brand_voice"
            m["customer_id"] = customerId
            m["brand_voice_id"] = brandVoiceId
            return m
        }
        try await LocalVectorStore.shared.upsert(
            customerId: customerId,
            entryType: "brand_voice",
            texts: chunks,
            metadatas: metas,
            db: db
        )
        return chunks.count
    }

    /// Embed and persist pricing rows.
    func ingestPricingRows(
        customerId: String,
        rows: [PricingRow],
        db: any Database
    ) async throws -> Int {
        let texts = rows.map { r in
            "Service: \(r.serviceType) | Price: USD \(r.priceUsd) | Won: \(r.won.map(String.init) ?? "unknown") | Notes: \(r.notes ?? "")"
        }
        let metas = rows.map { r -> [String: Any] in
            [
                "type": "pricing",
                "customer_id": customerId,
                "service_type": r.serviceType,
                "price_usd": r.priceUsd,
                "won": r.won as Any,
            ]
        }
        try await LocalVectorStore.shared.upsert(
            customerId: customerId,
            entryType: "pricing",
            texts: texts,
            metadatas: metas,
            db: db
        )
        return rows.count
    }

    // MARK: - Generate

    /// Full RAG pipeline: retrieve context → generate proposal with Claude.
    /// Automatically switches to cold-start when < 15 proposals indexed.
    /// BUG FIX: passes `industry` to cold-start so customers get the right
    /// template instead of always falling back to consulting defaults.
    func generateProposal(
        customerId: String,
        rfpText: String,
        proposalsIndexed: Int,
        industry: String?,
        db: any Database
    ) async throws -> String {
        let contextChunks: [VectorMatch]
        let coldStart: Bool

        if proposalsIndexed < ColdStart.shared.threshold {
            // Use cold-start templates with the customer's actual industry
            // Default to consulting template when industry is not specified — broadest coverage
        contextChunks = ColdStart.shared.getContext(industry: industry ?? "consulting")
            coldStart = true
        } else {
            contextChunks = try await LocalVectorStore.shared.query(
                customerId: customerId,
                queryText: rfpText,
                db: db
            )
            coldStart = false
        }

        return try await ClaudeService.shared.generateProposal(
            customerId: customerId,
            rfpText: rfpText,
            contextChunks: contextChunks,
            coldStart: coldStart,
            industry: industry
        )
    }
}

/// Lightweight pricing row for ingest requests.
struct PricingRow: Codable {
    let serviceType: String
    let priceUsd: Double
    let won: Bool?
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case serviceType = "service_type"
        case priceUsd    = "price_usd"
        case won, notes
    }
}

File: backend/services/rag_pipeline.py
"""
RAG (Retrieval-Augmented Generation) pipeline for Context-Mapper.

Handles:
1. Chunking long documents into overlapping 400-token chunks
2. Embedding and upserting to Pinecone (customer-isolated namespace)
3. Query-time retrieval with relevance filtering
4. Context-aware proposal generation via Claude Sonnet 4.6
"""
import uuid

from services.embeddings import embed_batch, embed_text
from services.pinecone_client import upsert_vectors, query_vectors, query_kb_vectors
from services.claude_client import generate_with_context

CHUNK_SIZE = 400        # tokens (~300 words)
CHUNK_OVERLAP = 80      # prevents cutting mid-sentence
RELEVANCE_THRESHOLD = 0.72  # minimum cosine similarity score


def chunk_text(text: str) -> list[str]:
    """
    Split text into overlapping word-boundary chunks.
    Overlap ensures semantic search finds relevant sections even at chunk boundaries.
    """
    words = text.split()
    chunks = []
    i = 0
    while i < len(words):
        chunk = " ".join(words[i : i + CHUNK_SIZE])
        chunks.append(chunk)
        i += CHUNK_SIZE - CHUNK_OVERLAP
    # Drop tiny trailing chunks (likely incomplete sentences)
    return [c for c in chunks if len(c.strip()) > 50]


async def ingest_proposal_text(
    customer_id: str,
    proposal_id: str,
    text: str,
    metadata: dict,
) -> int:
    """
    Chunk, embed, and upsert a full proposal document to the customer's namespace.

    metadata should include: title, client_name, outcome, value_usd, win_reason
    Returns the number of vectors upserted.
    """
    chunks = chunk_text(text)
    if not chunks:
        return 0

    embeddings = await embed_batch(chunks)
    vectors = [
        {
            "id": f"prop_{proposal_id}_chunk_{i}",
            "values": emb,
            "metadata": {
                "type": "proposal_chunk",
                "customer_id": customer_id,
                "proposal_id": proposal_id,
                "chunk_index": i,
                "text": chunks[i],
                **metadata,
            },
        }
        for i, emb in enumerate(embeddings)
    ]
    upsert_vectors(customer_id, vectors)
    return len(vectors)


async def ingest_brand_voice_text(
    customer_id: str,
    brand_voice_id: str,
    text: str,
    metadata: dict,
) -> int:
    """Embed and upsert brand voice examples to the customer's namespace."""
    chunks = chunk_text(text)
    if not chunks:
        return 0

    embeddings = await embed_batch(chunks)
    vectors = [
        {
            "id": f"bv_{brand_voice_id}_chunk_{i}",
            "values": emb,
            "metadata": {
                "type": "brand_voice",
                "customer_id": customer_id,
                "brand_voice_id": brand_voice_id,
                "chunk_index": i,
                "text": chunks[i],
                **metadata,
            },
        }
        for i, emb in enumerate(embeddings)
    ]
    upsert_vectors(customer_id, vectors)
    return len(vectors)


async def ingest_pricing_rows(customer_id: str, rows: list[dict]) -> int:
    """
    Embed pricing rows and upsert to the customer's namespace.
    Text format: "Service: {type} | Price: USD {price} | Won: {won} | Notes: {notes}"
    """
    texts = [
        f"Service: {r['service_type']} | Price: USD {r['price_usd']} | "
        f"Won: {r.get('won', 'unknown')} | Notes: {r.get('notes', '')}"
        for r in rows
    ]
    embeddings = await embed_batch(texts)
    vectors = [
        {
            "id": f"pricing_{customer_id}_{i}_{uuid.uuid4().hex[:8]}",
            "values": emb,
            "metadata": {
                "type": "pricing",
                "customer_id": customer_id,
                "service_type": rows[i]["service_type"],
                "price_usd": rows[i]["price_usd"],
                "won": rows[i].get("won"),
                "text": texts[i],
            },
        }
        for i, emb in enumerate(embeddings)
    ]
    upsert_vectors(customer_id, vectors)
    return len(vectors)


async def retrieve_context(
    customer_id: str,
    query: str,
    top_k: int = 8,
) -> list[dict]:
    """
    RAG retrieval: embed the query, find the most relevant chunks for this customer.
    Filters by relevance threshold (0.72) to avoid noise.
    Returns list of metadata dicts sorted by relevance score.
    """
    query_vector = await embed_text(query)
    matches = query_vectors(customer_id, query_vector, top_k=top_k)
    return [
        {
            "score": m.score,
            "text": m.metadata.get("text", ""),
            "type": m.metadata.get("type", ""),
            **m.metadata,
        }
        for m in matches
        if m.score > RELEVANCE_THRESHOLD
    ]


async def retrieve_kb_context(query: str, top_k: int = 5) -> list[dict]:
    """Retrieve relevant KB articles for support ticket deflection."""
    query_vector = await embed_text(query)
    matches = query_kb_vectors(query_vector, top_k=top_k)
    return [
        {
            "score": m.score,
            "text": m.metadata.get("text", ""),
            **m.metadata,
        }
        for m in matches
        if m.score > 0.65  # slightly lower threshold for KB (broader match)
    ]


async def generate_proposal_with_rag(
    customer_id: str,
    rfp_text: str,
    proposals_indexed: int,
) -> str:
    """
    Full RAG pipeline: retrieve context → generate proposal with Claude.
    Automatically switches to cold-start mode when < 15 proposals indexed.
    """
    from services.cold_start import COLD_START_THRESHOLD, get_cold_start_context

    if proposals_indexed < COLD_START_THRESHOLD:
        # Use industry template library for cold-start customers
        # customer industry needs to come from caller
        context_chunks = []
        result = await generate_with_context(
            customer_id=customer_id,
            rfp_text=rfp_text,
            context_chunks=context_chunks,
            cold_start=True,
        )
    else:
        context_chunks = await retrieve_context(customer_id, rfp_text)
        result = await generate_with_context(
            customer_id=customer_id,
            rfp_text=rfp_text,
            context_chunks=context_chunks,
            cold_start=False,
        )
    return result
